src/libs/elektra/keyname.c

```diff
@@ -502,176 +395,732 @@ static int keyGetNameNamespace (const char * name)
+	while (name - backslashes >= pathStart && *(name - backslashes) == '\\')
+	{
+		++backslashes;
+	}
+
+	if (backslashes % 2 != 0)
+	{
+		// dangling escape
+		return 0;
+	}
+
+	enum
+	{
+		PART_END,
+		DIGITS,
+		UNDERSCORES,
+		OTHER
+	} state = PART_END;
+
+	char prev = '\0';
+	size_t partLen = 0;
+	size_t digits = 0;
+	size_t underscores = 0;
+	size_t skip = 0;
+	while (name >= pathStart)
+	{
+		size_t trailingSlashes = 0;
+		if (state == PART_END)
+		{
+			// count trailing slashes (the last one might be escaped)
+			while (name - trailingSlashes >= pathStart && *(name - trailingSlashes) == '/')
+			{
+				++trailingSlashes;
+			}
+		}
+
+		// count upcomming backslashes
+		backslashes = 0;
+		size_t backslashOffset = trailingSlashes == 0 ? 1 : trailingSlashes;
+		while (name - backslashOffset - backslashes >= pathStart && *(name - backslashOffset - backslashes) == '\\')
+		{
+			++backslashes;
+		}
+
+		int escaped = backslashes % 2 != 0;
+
+		// the last trailing slash is actually escaped
+		if (trailingSlashes > 0 && escaped)
+		{
+			--trailingSlashes;
+		}
+
+		name -= trailingSlashes;
+		size -= trailingSlashes;
+		usize -= trailingSlashes;
+
+		char cur = *name;
+
+		switch (state)
+		{
+		case PART_END:
+			if (isdigit (cur))
+			{
+				digits = 1;
+				partLen = 1;
+				state = DIGITS;
+			}
+			else
+			{
+				partLen = 1;
+				state = OTHER;
+			}
+			break;
+		case DIGITS:
+			if (isdigit (cur))
+			{
+				++digits;
+			}
+			else
+			{
+				if (cur == '_')
+				{
+					underscores = 1;
+					state = UNDERSCORES;
+				}
+				else
+				{
+					state = OTHER;
+				}
+			}
+			break;
+		case UNDERSCORES:
+			if (cur == '_')
+			{
+				++underscores;
+				state = UNDERSCORES;
+			}
+			else
+			{
+				state = OTHER;
+			}
+			break;
+		case OTHER:
+			digits = 0;
+			underscores = 0;
+			break;
+		}
+
+		// move past cur
+		--name;
+
+		// move past upcomming backslashes
+		name -= backslashes;
+		if (skip == 0)
+		{
+			usize -= backslashes / 2;
+		}
+		partLen += backslashes;
+
+		if (escaped)
+		{
+			if (skip == 0)
+			{
+				--usize;
+			}
+
+			// uneven number of backslashes -> check escape sequence
+			if (backslashes == 1 && (name < pathStart || *name == '/'))
+			{
+				if (cur != '\\' && cur != '/' && cur != '.' && cur != '#' && cur != '%' && cur != '@')
+				{
+					// illegal escape sequence (at start of part)
+					return 0;
+				}
+			}
+			else
+			{
+				if (cur != '\\' && cur != '/')
+				{
+					// illegal escape sequence (in middle of part)
+					return 0;
+				}
+			}
+		}
+		else if (backslashes > 0)
+		{
+			// even number of backslashes -> cur is '\\' after moving past backslashes
+			cur = '\\';
+		}
+
+		int partStart = 0;
+		if (name < pathStart || *name == '/')
+		{
+			size_t count = 0;
+			while (name - 1 - count >= pathStart && *(name - 1 - count) == '\\')
+			{
+				++count;
+			}
+
+			if (count % 2 == 0)
+			{
+				partStart = 1;
+			}
+		}
+
+		if (state != PART_END && partStart)
```

Comment:
I can try, but honestly I don't think it will change anything. Personally, I think splitting this function will make maintaining it harder not easier. Other than extracting the bits before and after the loop, I don't see a good place to split the function. The result will be multiple functions with large numbers of arguments. These functions won't make any sense, if you don't understand the main function and you can't understand the main function without understanding every detail of the "helper" functions.

Another (IMO more important) point is that this function will likely have to change quite drastically before 1.0, if we want to implement any of the ideas related to restrictions on key names. To implement the change, we will essentially have to merge the functions back together, do the change and split it into new parts. In short, refactoring now seems a waste of time.

Reply:
I would rather love to see some functions in the middle part. If the functions get good names of what they do it should be a big improvement in any case. At the moment elektraKeyNameValidate is very hard to understand/review. keySetName is done very nicely :sparkling_heart:

Reply:
> I would rather love to see some functions in the middle part.

Me too, but I don't think there is an easy way to do that. The new function will be so tightly coupled to the existing one that I don't think there is any benefit. If anything it might make it more difficult to understand, because you need to keep track of two functions at the same time.

I will see what I can do, but I can't promise anything...

> At the moment elektraKeyNameValidate is very hard to understand/review.

That's why I added the Pyhton implementation and said that one is easier to read.